compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -1122,28 +1118,6 @@ private void updateClosureVars(BLangSimpleVarRef varRefExpr, BVarSymbol mapSymbo
         result = rewriteExpr(desugar.addConversionExprIfRequired(accessExpr, varRefExpr.type));
```

Comment:
Thinking out loud, the closure map is a `map<any|error>` right?

Looking at the comments and the logic, it seems like there' some legacy code (based on simple basic types vs ref types in the BVM) which may not be valid now.

So basically what I believe we've tried to do here is if it is a simple assignment like `x = 1;` where the ref we're working is on the LHS, we use `any|error` as the type of index-based access for simple basic types with an eventual cast on the RHS, whereas for ref types we're using the original types (which were compatible with the ref type impl). IIRC, we used to need this in the BVM based implementation (boxing), I'm not sure if we still need to do it in the front-end.

Where we have the closure map `map<any|error> $innerMap$1`.

```ballerina
x = 1;
```
becomes
```ballerina
$innerMap$1["x"] = <any> 1;
```

Then we have the access on the RHS, where in L1118 we've added a cast to the expected type. Basically, we access the relevant value from the `map<any|error>` and cast to the expected type.
```ballerina
int i = x;
```
becomes
```ballerina
int i = <int> $innerMap$1["x"] ;
```

So coming back to the problem we're trying to solve,

```ballerina
    var fn = function () {
        fb.i = 3;
    };
```
it seems like even for this, it goes in the lhsVar path, without the cast. We seem to be returning `$innerMap$1["fb"]` instead of `<Foo|Bar> $innerMap$1["fb"]`.

IMO we can do one of the following.

1. Semantically, a map (`map<T>`) access on the RHS returns `T?` so if someone writes this code in Ballerina, the type of access expr will be `T` on the LHS, but `T?` on the RHS. With the current changes, we're changing this for both LHS and RHS to `T`, but I guess i's OK since we know that value is always present in the map. If we are going ahead with this though let's check
a) if this works as expected for simple basic types (int, byte, float, etc.)
b) if we need the `desugar.addConversionExprIfRequired(accessExpr, varRefExpr.type)` anymore since the access expression's type is now varRefExpr.type

or

2. we need to figure out if we can add the cast on the LHS side also if the current var ref is part of a compound expression.